### PR TITLE
refresh bitops module

### PR DESCRIPTION
**Pull request by arnetheduck**
* switch from `proc {.nosideeffect.}` to `func`
* deprecate compile time -d switch in favor of API argument
  * compile time switches don't work with larger projects where requirements differ between modules
* deprecate bit operations for signed ints
  * it doesn't quite make sense, and for the cases where it does make sense, an explicit cast seems reasonable as "warning"
* add some docs

this patch also defaults bitops such in a way that favors no undefined states by checking for 0 - in 2018, modern cpu's and compilers optimize this to a single instruction (ie `LZCNT`)

~lastly, it removes `{.inline.}` by the same argument - compilers will optimize it away anyway and it just adds noise in the declaration - when performance matters, optimizing backends should inline them - I'm 50/50 on this one though because unfortunately some of these optimizations require lto for which nim doesn't have support yet..~
edit: readded inlines - compiler not ready for dropping this ugly noise just yet

in this patch, names have been changed so as to use `one` and `zero` instead of `set` and `zero`. another refactoring with regards to names would be to remove `bits` from everywhere - this is unnecessary repetition from the `bitops` module, but perhaps plays poorly with the nim culture of not namespacing things

### CODE REVIEW

File: lib/pure/bitops.nim [hunk 1]
```diff
@@ -163,174 +164,208 @@ elif useICC_builtins:
...
-      when sizeof(x) <= 4: result = parity_impl(x.uint32)
-      else:                result = parity_impl(x.uint64)
+      when sizeof(x) <= 4: parityImpl(x.uint32)
+      else:                parityImpl(x.uint64)
+
+func parityBits*(x: SomeSignedInt): int {.deprecated.} =
+  parityBits(x.uint64)
 
-proc firstSetBit*(x: SomeInteger): int {.inline, nosideeffect.} =
+func firstOneBit*(x: SomeUnsignedInt, maybeZero = true): int =
```
Comment (mratsim) [inline lib/pure/bitops.nim]: This should be a static bool. I don't se any use-case where having it runtime defined is useful.

File: lib/pure/bitops.nim [hunk 2]
```diff
@@ -163,174 +164,208 @@ elif useICC_builtins:
...
+  ## Example:
+  ## doAssert firstOneBit(0b00000010'u8) == 2
+  ##
   when nimvm:
-    when noUndefined:
-      if x == 0:
-        return 0
-    when sizeof(x) <= 4: result = firstSetBit_nim(x.uint32)
-    else:                result = firstSetBit_nim(x.uint64)
+    if maybeZero and x == 0: 0
```
Comment (mratsim) [inline lib/pure/bitops.nim]: With a static bool, GCC/LLVM can optimize this + the following logic into LZCNT

File: lib/pure/bitops.nim [hunk 3]
```diff
@@ -163,174 +164,208 @@ elif useICC_builtins:
...
-    when sizeof(x) <= 4: result = fastlog2_nim(x.uint32)
-    else:                result = fastlog2_nim(x.uint64)
+      if maybeZero and x == 0: 0
+      elif sizeof(x) <= 4: firstOneBitNim(x.uint32)
+      else:                firstOneBitNim(x.uint64)
+
+func firstSetBit*(x: SomeInteger): int {.deprecated: "firstOneBit".} =
+  firstOneBit(x.uint64, noUndefined)
+
+func fastLog2Bit*(x: SomeUnsignedInt, maybeZero = true): int =
```
Comment (mratsim) [inline lib/pure/bitops.nim]: static bool

File: lib/pure/bitops.nim [hunk 4]
```diff
@@ -163,174 +164,208 @@ elif useICC_builtins:
...
+        else:                fastlog2Nim(x.uint64)
+      else:
+        when sizeof(x) <= 4: fastlog2Nim(x.uint32)
+        else:                fastlog2Nim(x.uint64)
 
-proc countLeadingZeroBits*(x: SomeInteger): int {.inline, nosideeffect.} =
+func fastlog2*(x: SomeInteger): int {.deprecated: "fastLog2Bit".} =
+  fastLog2Bit(x.uint64, noUndefined)
+
+func leadingZeroBits*(x: SomeInteger, maybeZero = true): int =
```
Comment (mratsim) [inline lib/pure/bitops.nim]: static bool

File: lib/pure/bitops.nim (hunk 1 above)
Comment (mratsim) [reply lib/pure/bitops.nim]: I would keep the inline as well here. In a worse case scenario the operations are:

on x86:
BSR + XOR + SUB (if the compiler doesn't cancel out the sub and xor)

With a non-inline function you would have to add push, pop, ret and doubling the cost with no code size advantage.

File: lib/pure/bitops.nim (hunk 4 above)
Comment (mratsim) [reply lib/pure/bitops.nim]: and inline as well

File: lib/pure/bitops.nim (hunk 1 above)
Comment (arnetheduck) [reply lib/pure/bitops.nim]: `static[bool]`: there's no semantic reason for it and it makes the declaration a lot more ugly - requires users of the api to waste more effort on understanding what it means. compilers nowadays do interprocedural constant propagation, ie https://kristerw.blogspot.com/2017/05/interprocedural-optimization-in-gcc.html. honestly, the best I can think of is actually to remove the parameter completely

`inline`: as noted in intro, a feature that adds visual noise and where the compiler is generally in a better spot to make the right tradeoff, based on contextual information (target instruction set, size-vs-speed flags, other operations near the call, constant folding, subexpression eval etc etc.. the only reason to keep it really is that nim doesn't use lto or amalgamation by default meaning that the way the c backend is setup, inlining won't happen across module boundaries which is a shame, also for constant propagation

File: lib/pure/bitops.nim (hunk 2 above)
Comment (arnetheduck) [reply lib/pure/bitops.nim]: it can also optimize it without a static bool

**Comment by arnetheduck** [comment]
since this new and nice feature was added to the language, doesn't it make more sense to slowly (or quickly) start getting rid of {.nosideeffects.}? half of it can be done with a search-replace, and then the consistency argument would flip the other way. it's really less noisy than a pragma

**Review by krux02 — CHANGES_REQUESTED** [review]
This branch now has conflicts since bitops needed some special treatment on 32 bit platforms. So I can't merge it anymore.

The reason this PR hasn't been touched or merged yet is, it does too many things at once. Ideally each of your points should have been it's own PR. Small PRs are faster to process than this one.

Most of the diff that hids the important details comes from changes proc to func, and I am sorry, it is out of the discussion if we do thing refactoring now, because we don't. There is no ``func`` used in the compiler or standard library at all. These changes need to be reverted

For the deprecation of the compile time switch, good work. I am totally on your side. This should not have been a compile time switch in the first place.

> deprecate bit operations for signed ints

Here im am only generally on your side. These features should have been implemented for unsigned integers only. But I also don't see the need for a change that triggers annoying deprecation warnings either. So please don't do it.
  
> add some docs

Documentation improvements are always good.

**Comment by arnetheduck** [comment]
ok, maybe some other time, don't have the focus to take care of this now.

**Pull request closed without merging**